docs/source/type_narrowing.rst

```diff
@@ -0,0 +1,324 @@
+
+    # Incorrect: returns True for some non-ints
+    def bad_typeis2(x: object) -> TypeIs[int]:
+        return isinstance(x, (int, float))
+
+This function demonstrates some errors that can occur when using a poorly written
+``TypeIs`` function. These errors are not detected by type checkers::
+
+    def caller(x: int | str, y: int | float) -> None:
+        if bad_typeis1(x):  # narrowed to int
+            print(x + 1)
+        else:  # narrowed to str (incorrectly)
+            print("Hello " + x)  # runtime error if x is a negative int
+
+        if bad_typeis2(y):  # narrowed to int
+            # Because of the incorrect TypeIs, this branch is taken at runtime if
+            # y is a float.
+            print(y.bit_count())  # runtime error: this method exists only on int, not float
+        else:  # narrowed to float (though never executed at runtime)
+            pass
+
+Here is an example of a correct ``TypeIs`` function for a more complicated type::
+
+    from typing import TypedDict, TypeIs
+
+    class Point(TypedDict):
+        x: int
+        y: int
+
+    def is_point(x: object) -> TypeIs[Point]:
+        return (
+            isinstance(x, dict)
+            and all(isinstance(key, str) for key in x)
+            and "x" in x
+            and "y" in x
+            and isinstance(x["x"], int)
+            and isinstance(x["y"], int)
+        )
+
+.. _`guide-type-narrowing-typeis-typeguard`:
+
+``TypeIs`` and ``TypeGuard``
+----------------------------
+
+:py:data:`typing.TypeIs` and :py:data:`typing.TypeGuard` are both tools for narrowing the type of a variable
+based on a user-defined function. Both can be used to annotate functions that take an
+argument and return a boolean depending on whether the input argument is compatible with
+the narrowed type. These function can then be used in ``if`` checks to narrow the type
+of a variable.
+
+``TypeIs`` usually has the most intuitive behavior, but it
+introduces more restrictions. ``TypeGuard`` is the right tool to use if:
+
+* You want to narrow to a type that is not compatible with the input type, for example
+  from ``list[object]`` to ``list[int]``.  ``TypeIs`` only allows narrowing between
+  compatible types.
+* Your function does not return ``True`` for all input values that are compatible with
+  the narrowed type. For example, you could have a ``TypeGuard[int]`` that returns ``True``
+  only for positive integers.
+
+``TypeIs`` and ``TypeGuard`` differ in the following ways:
+
+* ``TypeIs`` requires the narrowed type to be a subtype of the input type, while
+  ``TypeGuard`` does not.
+* When a ``TypeGuard`` function returns ``True``, type checkers narrow the type of the
+  variable to exactly the ``TypeGuard`` type. When a ``TypeIs`` function returns ``True``,
+  type checkers can infer a more precise type combining the previously known type of the
+  variable with the ``TypeIs`` type. (Technically, this is known as an intersection type.)
```

Comment:
Not sure "technically" adds much here: the whole document is technical :)

```suggestion
  variable with the ``TypeIs`` type. (This is known as an "intersection type".)
```

You could possibly say

```suggestion
  variable with the ``TypeIs`` type. (In type theory, this is known as an
  "intersection type".)
```

Reply:
Thanks, I think the first suggestion is best (when in doubt, I like to go with the less verbose version).